src/lib/comms/sol-http-common.c

```diff
@@ -123,3 +120,453 @@ sol_http_escape_string(char **escaped, const char *value)
+{
+    struct sol_str_slice query, *token;
+    char *i = partial_uri, *sep;
+    uint16_t j;
+    struct sol_vector tokens;
+    int r = 0;
+
+    if (!*i)
+        goto exit;
+
+    if (*i != '?')
+        goto exit;
+
+    query.data = ++i;
+
+    for (; *i && *i != '#'; i++) {
+        if (*i != '/' && *i != '?' && !_is_pchar(*i)) {
+            SOL_ERR("Invalid character while parsing the query. "
+                "Found: %c", *i);
+            return -EINVAL;
+        }
+    }
+    query.len = i - query.data;
+
+    tokens = sol_util_str_split(query, "&", 0);
+
+    SOL_VECTOR_FOREACH_IDX (&tokens, token, j) {
+        struct sol_http_param_value param;
+        sep = memchr(token->data, '=', token->len);
+        if (!sep) {
+            SOL_ERR("The param: %.*s has no \"=\" separator",
+                SOL_STR_SLICE_PRINT(*token));
+            r = -EINVAL;
+            break;
+        }
+
+        param.value.key_value.key.data = token->data;
+        param.value.key_value.key.len = sep - token->data;
+
+        param.value.key_value.value.data = sep + 1;
+        param.value.key_value.value.len = token->len -
+            param.value.key_value.key.len - 1;
+
+        SOL_DBG("Query key: *%.*s* Query value: *%.*s*",
+            SOL_STR_SLICE_PRINT(param.value.key_value.key),
+            SOL_STR_SLICE_PRINT(param.value.key_value.value));
+
+        param.type = 0;
+
+        if (!sol_http_param_add(query_params, param)) {
+            SOL_ERR("Could not alloc the param %.*s : %.*s",
+                SOL_STR_SLICE_PRINT(param.value.key_value.key),
+                SOL_STR_SLICE_PRINT(param.value.key_value.value));
+            r = -ENOMEM;
+            break;
+        }
+    }
+
+    sol_vector_clear(&tokens);
+exit:
+    *next = i;
+    return r;
+}
+
+static int
+_get_fragment(char *partial_uri, struct sol_str_slice *fragment)
+{
+    struct sol_str_slice f = SOL_STR_SLICE_EMPTY;
+    char *i = partial_uri;
+
+    if (!*i)
+        goto exit;
+
+    if (*i != '#') {
+        SOL_ERR("A \"#\" is required in order to identify the fragment");
+        return -EINVAL;
+    }
+
+    f.data = ++i;
+    for (; *i; i++) {
+        if (*i != '/' && *i != '?' && !_is_pchar(*i)) {
+            SOL_ERR("Invalid character while parsing the fragement. "
+                "Found: %c", *i);
+            return -EINVAL;
+        }
+    }
+    f.len = i - f.data;
+
+exit:
+    SOL_DBG("URI Fragment: %.*s", SOL_STR_SLICE_PRINT(f));
+    if (fragment)
+        *fragment = f;
+    return 0;
+}
+
+SOL_API int
+sol_http_split_uri(const char *uri, struct sol_str_slice *protocol,
+    struct sol_str_slice *server, struct sol_str_slice *path,
+    struct sol_str_slice *fragment, struct sol_http_param *query_params,
+    int *port)
+{
+    char *next;
+    int r;
+
+    SOL_NULL_CHECK(uri, -EINVAL);
+
+    SOL_DBG("Splitting URI: %s", uri);
+
+    r = _get_protocol(uri, protocol, &next);
+    SOL_INT_CHECK(r, < 0, r);
+    r = _get_server(next, server, &next);
```

Comment:
Well, we made a naive parser ;).
I will improve it, so it can handle all the possible URI formats.